app/uk/gov/hmrc/apisubscriptionfields/controller/ApiSubscriptionFieldsController.scala

```diff
@@ -17,25 +17,65 @@
-
-	def id() = Action.async { implicit request =>
-    /*
-     TODO:
-       parse `ApiSubscriptionRequest`
-       - [application_id]_[api_context]_[api_context] is the id of the mongo record, but decide good separators
-       - generate UUID (the header to be passed in the incoming requests)
-       - save record to mongo
-       */
-    val subscriptionFieldsId = UUID.randomUUID()
-    Future.successful(Created(s"$subscriptionFieldsId"))
+@Singleton
+class ApiSubscriptionFieldsController @Inject() (service: SubscriptionFieldsService) extends CommonController {
+
+  import JsonFormatters._
+
+  private val NOT_FOUND_RESPONSE = NotFound(JsErrorResponse(SUBSCRIPTION_FIELDS_ID_NOT_FOUND, "Subscription Fields were not found"))
+
+  def id(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[AnyContent] = Action.async { implicit request =>
+    service.get(SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion))) map {
+      case Some(subscriptionFields) => Ok(Json.toJson(SubscriptionFieldsIdResponse(subscriptionFields.fieldsId)))
+      case None => NOT_FOUND_RESPONSE
+    } recover recovery
   }
-}
 
-object ApiSubscriptionFieldsController extends ApiSubscriptionFieldsController
+  def getSubscriptionFields(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[AnyContent] = Action.async { implicit request =>
+    service.get(SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion))) map {
+      case Some(subscriptionFields) => Ok(Json.toJson(subscriptionFields))
+      case None => NOT_FOUND_RESPONSE
+    } recover recovery
+  }
+
+
+  def getSubscriptionFieldsByFieldsId(rawFieldsId: UUID): Action[AnyContent] = Action.async { implicit request =>
+    service.get(SubscriptionFieldsId(rawFieldsId)) map {
+      case Some(subscriptionFields) => Ok(Json.toJson(subscriptionFields))
+      case None => NOT_FOUND_RESPONSE
+    } recover recovery
+  }
+
+
+  def upsertSubscriptionFields(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[JsValue] = Action.async(parse.json) { implicit request =>
+      withJsonBody[SubscriptionFieldsRequest] { payload =>
+        service.upsert(SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion)), payload.fields) map {
+          case (response, true) => Created(Json.toJson(response))
+          case (response, false) => Ok(Json.toJson(response))
+        } recover {
+          case e: Exception => BadRequest(e.getMessage)
+        }
+      } recover recovery
+    }
+
+
+  def deleteSubscriptionFields(rawAppId: UUID, rawApiContext: String, rawApiVersion: String): Action[AnyContent] = Action.async { implicit request =>
+    val identifier = SubscriptionIdentifier(AppId(rawAppId), ApiContext(rawApiContext), ApiVersion(rawApiVersion))
+    service.get(identifier) flatMap {
+      case None => Future.successful(NOT_FOUND_RESPONSE)
```

Comment:
if you change the service to return `Future[Boolean]`, then you will map:
- `true` to `NoContent `
- `false` to `NotFound`

Reply:
Done